torch/testing/asserts.py

```diff
@@ -0,0 +1,389 @@
+
+    _assert_attributes_equal(a, b, check_device=check_device, check_dtype=check_dtype, check_stride=check_stride)
+
+    if a.device != b.device:
+        a = a.cpu()
+        b = b.cpu()
+
+    if a.dtype != b.dtype:
+        dtype = _get_comparison_dtype(a, b)
+        a = a.to(dtype)
+        b = b.to(dtype)
+
+    values_asserter(a, b)
+
+
+@_hide_internal_traceback_pytest
+def assert_tensors_equal(
+    a: torch.Tensor,
+    b: torch.Tensor,
+    /,
+    *,
+    warn_floating_point_dtype: bool = True,
+    check_device: bool = True,
+    check_dtype: bool = True,
+    check_stride: bool = True,
+) -> None:
+    """Asserts that the values of two tensors are bitwise equal.
+
+    Optionally, checks that some attributes of both tensors are equal.
+
+    Args:
+        a (torch.Tensor): First tensor.
+        b (torch.Tensor): Second tensor.
+        warn_floating_point_dtype (bool): If ``True`` (default), emit a warning if :attr:`a` or :attr:`b` is a
+            floating point tensor.
+        check_device (bool): If ``True`` (default), asserts that both :attr:`a` and :attr:`b` live in the same
+            :attr:`~torch.Tensor.device` memory. If this check is disabled **and** :attr:`a` and :attr:`b` do not live
+            in the same memory :attr:`~torch.Tensor.device`, they are moved to the CPU memory.
+        check_dtype (bool): If ``True`` (default), asserts that both :attr:`a` and :attr:`b` have the same
+            :attr:`~torch.Tensor.dtype`. If this check is disabled **and** :attr:`a` and :attr:`b` do not have the same
+            :attr:`~torch.Tensor.dtype`, the comparison :attr:`~torch.Tensor.dtype` is determined by
+            :func:`torch.promote_types`.
+        check_stride (bool): If ``True`` (default), asserts that both :attr:`a` and :attr:`b` have the same
+            :meth:`~torch.Tensor.stride`.
+
+    Raises:
+        AssertionError: If :attr:`a` and :attr:`b` do not have the same :attr:`~torch.Tensor.shape`.
+        AssertionError: If :attr:`check_device`, but :attr:`a` and :attr:`b` do not live in the same
+            :attr:`~torch.Tensor.device` memory.
+        AssertionError: If :attr:`check_dtype`, but :attr:`a` and :attr:`b` do not have the same
+            :attr:`~torch.Tensor.dtype`.
+        AssertionError: If :attr:`check_stride`, but :attr:`a` and :attr:`b` do not have the same
+            :meth:`~torch.Tensor.stride`.
+        AssertionError: If the values of :attr:`a` and :attr:`b` are not bitwise equal.
+        UsageError: If :attr:`a` or :attr:`b` is complex, quantized, or sparse. This is a temporary restriction and
+            will be relaxed in the future.
+
+    .. seealso::
+
+        To compare tensors for value proximity, :func:`assert_tensors_allclose` can be used.
+    """
+    _assert_are_tensors(a, b)
+
+    is_floating_point = a.dtype.is_floating_point or b.dtype.is_floating_point
+    if is_floating_point and warn_floating_point_dtype:
```

Comment:
This is fine to include but I'm not sure this warning does much. It's hard to imagine an engineer writing tests for a project using PyTorch but who also doesn't appreciate that floating point computations are imprecise.

Reply:
I've included this since our current method is named `assertEqual` but does closeness checks for floating point tensors by default. I simply wanted to avoid surprises for someone that without further thoughts uses `assert_tensors_equal()` as a drop-in replacement. Thoughts?